src/plugins/intel_npu/src/backend/src/zero_infer_request.cpp

```diff
@@ -307,13 +213,16 @@ void ZeroInferRequest::infer_async() {
     _logger.debug("InferRequest::infer_async started");
     OV_ITT_SCOPED_TASK(itt::domains::LevelZeroBackend, "infer_async");
 
-    for (const auto& name : _inputAndStateInputNames) {
-        auto& inputTensor = _allTensors.at(name);
-        const auto& wrapperInputTensor = _copyAllTensors.at(name);
+    size_t inputIndex = 0;
+    for (const std::shared_ptr<ov::ITensor>& inputTensor : _inputTensors) {
+        const std::shared_ptr<ov::ITensor>& copyInputTensor = _copyInputTensors.at(inputIndex);
 
-        if (isShapeTensorName(name)) {
-            const auto actualTensorName = name.substr(SHAPE_TENSOR_PREFIX.size());
-            const auto& inputDims = _allTensors.at(actualTensorName)->get_shape();
+        const IODescriptor inputDescriptor = _metadata.inputs.at(inputIndex);
+        if (inputDescriptor.isShapeTensor) {
+            OPENVINO_ASSERT(inputDescriptor.relatedDescriptorIndex.has_value(),
+                            "The link between the dynamic tensor and its shape tensor is missing, entry name: ",
+                            inputDescriptor.nameFromCompiler);
```

Comment:
See bindRelatedDescriptors() inside icompiler.hpp for more details on how these links are established.